packages/core/src/components/category-filter/category-filter.tsx

```diff
@@ -154,49 +164,59 @@ export class CategoryFilter {
   @Event() filterChanged!: EventEmitter<FilterState>;
 
   get dropdown() {
-    return this.hostElement.shadowRoot.querySelector('ix-dropdown');
+    return this.hostElement.shadowRoot?.querySelector('ix-dropdown');
   }
 
   @Watch('filterState')
-  watchFilterState(newValue) {
+  watchFilterState(newValue: FilterState) {
     this.setFilterState(newValue);
   }
 
-  componentDidLoad() {
-    if (this.filterState !== undefined) {
-      setTimeout(() => this.setFilterState(this.filterState));
+  private preventDefault(e: Event) {
+    e.preventDefault();
+  }
+
+  private onFocusIn() {
+    this.hasFocus = true;
+  }
+
+  private onFocusOut() {
+    this.hasFocus = false;
+  }
+
+  private onInput() {
+    this.inputValue = this.textInput?.current?.value || '';
+    const inputState = new InputState(this.inputValue, this.category);
+    this.inputChanged.emit(inputState);
+
+    if (!this.dropdown?.show) {
+      this.openDropdown();
     }
+  }
 
-    this.hostElement?.addEventListener(
-      'keydown',
-      this.handleFormElementKeyDown.bind(this)
-    );
+  componentDidLoad() {
+    setTimeout(() => {
+      if (this.filterState !== undefined) {
+        this.setFilterState(this.filterState);
+      }
+    });
 
-    this.formElement?.addEventListener('submit', (e) => e.preventDefault());
+    this.hostElement?.addEventListener('keydown', this.formElementKeyDownBind);
+    this.formElement?.addEventListener('submit', this.preventDefaultBind);
 
-    if (this.textInput == null) {
+    if (this.textInput?.current == null) {
       console.warn(
         'ix-category-filter - unable to add event listeners to native input element'
       );
       return;
     }
 
-    this.textInput.addEventListener('focusin', () => {
-      this.hasFocus = true;
-    });
-    this.textInput.addEventListener('focusout', () => (this.hasFocus = false));
-    this.textInput.addEventListener('input', () => {
-      this.inputValue = this.textInput.value;
-      const inputState = new InputState(this.inputValue, this.category);
-      this.inputChanged.emit(inputState);
-
-      if (!this.dropdown.show) {
-        this.openDropdown();
-      }
-    });
-    this.textInput.addEventListener(
+    this.textInput.current.addEventListener('focusin', this.onFocusInBind);
+    this.textInput.current.addEventListener('focusout', this.onFocusOutBind);
+    this.textInput.current.addEventListener('input', this.onInputBind);
+    this.textInput.current.addEventListener(
       'keydown',
-      this.handleInputElementKeyDown.bind(this)
+      this.inputElementKeyDownBind
```

Comment:
`bind` make the code really ugly. This should be refactored to a dispose pattern.

Make the addDisposableEventListener(https://github.com/siemens/ix/blob/main/packages/core/src/components/dropdown/dropdown-controller.ts#L216-L240) to a `util` file and use the function and prevent the bind problem.